core/drlibc/drlibc_x86.asm

```diff
@@ -762,46 +762,72 @@ GLOBAL_LABEL(load_dynamo_failure:)
  * signature:
  *   thread_id_t dynamorio_clone(uint flags, byte *newsp, void *ptid, void *tls,
  *                               void *ctid, void (*func)(void))
+ * i#6514: If newsp is NULL then that tells the kernel to give the child the
+ * same value for SP as the parent.
  */
         DECLARE_FUNC(dynamorio_clone)
 GLOBAL_LABEL(dynamorio_clone:)
-        /* save func for use post-syscall on the newsp.
-         * when using clone_record_t we have 4 slots we can clobber.
+        /* Save func for use post-syscall on the newsp.
+         * This is tricky because we have to handle the case of newsp == NULL.
          */
 # ifdef X64
-        sub      ARG2, ARG_SZ
-        mov      [ARG2], ARG6 /* func is now on TOS of newsp */
-        /* all args are already in syscall registers */
+        /* The syscall preserves all registers except rax, rcx, r11. */
+        sub      REG_XSP, FRAME_ALIGNMENT
+        mov      [REG_XSP], r15
+        mov      r15, ARG6 /* func is now in r15 */
+        /* All args are already in syscall registers, except for rcx. */
         mov      r10, rcx
         mov      REG_XAX, SYS_clone
         syscall
 # else
-        mov      REG_XAX, ARG6
-        mov      REG_XCX, ARG2
-        sub      REG_XCX, ARG_SZ
-        mov      [REG_XCX], REG_XAX /* func is now on TOS of newsp */
-        mov      REG_XDX, ARG3
-        /* preserve callee-saved regs */
+        /* Fetch some args we need before we modify XSP and ARGn is no
+         * longer usable.
+         */
+        mov      REG_XCX, ARG2 /* newsp */
+        mov      REG_XDX, ARG3 /* ptid */
+        mov      REG_XAX, ARG6 /* func */
+        /* Preserve callee-saved regs. */
         push     REG_XBX
         push     REG_XSI
         push     REG_XDI
-        /* now can't use ARG* since xsp modified by pushes */
-        mov      REG_XBX, DWORD [4*ARG_SZ + REG_XSP] /* ARG1 + 3 pushes */
-        mov      REG_XSI, DWORD [7*ARG_SZ + REG_XSP] /* ARG4 + 3 pushes */
-        mov      REG_XDI, DWORD [8*ARG_SZ + REG_XSP] /* ARG5 + 3 pushes */
+        /* Save space for func for newsp == NULL case. */
+        sub      REG_XSP, ARG_SZ
+        /* i#6514: Save func on the child's stack. Remember that if newsp is
+         * NULL then the child's stack is our stack.
+         */
+        and      REG_XCX, HEX(fffffff0) /* Ensure child xsp is aligned. */
+        jnz      stack_not_null
+        lea      REG_XCX, [ARG_SZ + REG_XSP] /* Add ARG_SZ back for the sub. */
```

Comment:
Line 793 says you did the SUB there for this ecx==0 case, and here it says you're undoing a SUB which at first seems to be the line 794 SUB and so it all seems like nops and is confusing.  Really this comment refers to the line 802 SUB, right?  Please add "the sub *below*" or something.

Reply:
Simplified by just always pushing func onto the parent's stack, even if newsp != NULL.